packages/util/src/errors.ts

```diff
@@ -54,115 +54,95 @@
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+export class FirebaseError extends Error {
+  readonly name = ERROR_NAME;
 
   constructor(
-    private service: string,
-    private serviceName: string,
-    private errors: ErrorList<T>
+    readonly code: string,
+    message: string,
+    readonly data: ErrorData = {}
   ) {
-    // empty
-  }
+    super(message);
+
+    // Fix For ES5
+    // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, FirebaseError.prototype);
 
-  create(code: T, data?: { [prop: string]: StringLike }): FirebaseError {
-    if (data === undefined) {
-      data = {};
+    // Maintains proper stack trace for where our error was thrown.
+    // Only available on V8.
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, ErrorFactory.prototype.create);
     }
+  }
+}
 
-    let template = this.errors[code as string];
+export class ErrorFactory<ErrorCode extends string> {
+  constructor(
+    private readonly service: string,
+    private readonly serviceName: string,
+    private readonly errors: ErrorList<ErrorCode>
+  ) {}
 
-    let fullCode = this.service + '/' + code;
-    let message: string;
+  create(code: ErrorCode, data: ErrorData = {}): FirebaseError {
+    const fullCode = `${this.service}/${code}`;
+    const template = this.errors[code];
 
-    if (template === undefined) {
-      message = 'Error';
-    } else {
-      message = template.replace(this.pattern, (match, key) => {
-        let value = data![key];
-        return value !== undefined ? value.toString() : '<' + key + '?>';
-      });
-    }
+    const message = template ? replaceTemplate(template, data) : 'Error';
 
     // Service: Error message (service/code).
-    message = this.serviceName + ': ' + message + ' (' + fullCode + ').';
-    let err = new FirebaseError(fullCode, message);
-
-    // Populate the Error object with message parts for programmatic
-    // accesses (e.g., e.file).
-    for (let prop in data) {
-      if (!data.hasOwnProperty(prop) || prop.slice(-1) === '_') {
-        continue;
+    const fullMessage = `${this.serviceName}: ${message} (${fullCode}).`;
+
+    // Keys with an underscore at the end of their name are not included in
+    // error.data for some reason.
+    const filteredData: ErrorData = {};
+    // TODO: Replace with Object.entries when lib is updated to es2017.
+    for (const key of Object.keys(data)) {
+      if (key.slice(-1) !== '_') {
+        filteredData[key] = data[key];
       }
-      (err as any)[prop] = data[prop];
     }
-
-    return err;
+    return new FirebaseError(fullCode, fullMessage, filteredData);
   }
 }
+
+function replaceTemplate(template: string, data: ErrorData): string {
+  return template.replace(PATTERN, (_, key) => {
+    let value = data != null ? data[key] : undefined;
```

Comment:
I'm not sure who you mean by "people new to JS". Is it other Google engineers, or outside contributors?

If it's Google engineers, this is something that is explicitly allowed in [the style guide](http://go/tsstyle#equality-checks), which they should read if they're new to TS. If it's outside contributors, I don't think we should focus on making our code easy to understand for people who are new to JS. We're not teaching JS, we're building a product. Plus, I'd say `==` equality and it's quirks is one of JS fundamentals, and it's really easy to find tables like [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Loose_equality_using).

Anyway, I switched to `!= null` because most of our SDKs (including util) don't have `strictNullChecks` on. That means stuff can be `null` without any compiler errors. Plus it's the same amount of code, and it's (marginally) safer, so I thought why not? :slightly_smiling_face:

Looking at this again, I guess it's very unlikely that someone would pass in `null` explicitly for data, and if so, it would fail a few lines later anyway, so I'm removing that check. I'll keep `value != null` though since I do need to check there.

Reply:
I disagree with the style guide, but it's not a big deal and I'm okay with being consistent with it.
But I will state my counter points anyway :)

1. My original point - Using `!==` makes it easier to understand. Removing the need/confusion of dealing with javascript quirks is always a good thing.
2. Sometimes you might just want to check `null`, but not `undefined`. When I review the code, if you used `!=`, I will always ask the question - Do you mean to check both `null` and `undefined`, or  you actually want to only check null, but made a mistake. Being explicit can eliminate this confusion.

In this case, I can actually infer that it's the correct thing to check both `null` and `undefined`, so using `!=` is probably fine here, but it could be harder if there are more complicated code to read.

Reply:
There are some people who write JavaScript and TypeScript on a daily basis and don't know the quirks of loose equality comparison. I'm a contributor to this library. I'd honestly find it difficult to recall what it does exactly.

I know many people who don't use it for long years now, as a principle. These people, including myself, avoid it like the plague, the same way as `any` in TypeScript.

Is the following reasonable? `'\n  123  \t' == 123 // true` 😋

Reply:
@Feiyang1 Honestly, I disagree with the style guide too sometimes, like it allows single line `if` statements without `{}` brackets, which sounds very wrong to me :smile:. I love consistency though, and overall it's a pretty good guide.

I agree with using `=== undefined` (or `=== null`) if you don't need to check both BTW, I have been doing that in my recent SDK code as well. But I'd rather play it safe in packages where the `strict` flag isn't turned on, and I prefer doing `== null` instead of `=== null || === undefined`, maybe because as I'm already used to it, it doesn't look confusing to me anymore.

@merlinnot That looks horrible :smile:.

I didn't mean that remembering the quirks is fundamental, I actually did horribly [in this game](https://eqeq.js.org/) a few months ago, most likely because I almost always avoid `==`, just like you mentioned. I meant that almost everybody knows that `==` is weird, and if you know that, you can just look up what `== null` does in a few seconds. And doing `== null` to check both `null` and `undefined` is the only exception to the `==` ban in our style guide, so it's easy enough to remember after you see it a few times.

And `any` is the worst indeed :smile:.